## Shopping Cart Analysis

### Introduction
Data analysis of an e-commerce consists in accumulating data from an online store, extracting trends and key informations, and communicating results to help decision-makers to improve and develop the store. For example, one may identify as relevant data details on sales or personal informations on customers.

This notebook aims to present the analysis of a synthetic dataset for e-commerce shopping using PostgreSQL. 

**Dataset**\
This dataset is provided by Ruchi Bhatia (Carnegie Mellon University) on Kaggle. It consists of four CSV files:
 - *customers.csv*: personal informations on customers such as its name, gender, age, address, etc.
 - *orders.csv*: data on the orders such as the customer, the payment, the order and delivery dates
 - *products.csv*: informations on products such as its name, size, colour, price, etc.
 - *sales.csv*: informations on sales such as the concerned order and product, the price, the quantity, etc. 
 
<u>URL:</u> https://www.kaggle.com/datasets/ruchi798/shopping-cart-database

### Database initialization 

First, let's load the <code>ipython-sql</code> library to enable the use of SQL magic functions and connect to the PostgreSQL database named <code>shopping_cart</code>. 

In [1]:
%load_ext sql

In [2]:
%sql  postgresql://postgres:***@localhost:5433/shopping_cart

The <code>shopping_cart</code> database is filled by data extracted from CSV files. Four tables are created, i.e., <code>customers</code>, <code>orders</code>, <code>products</code> and <code>sales</code>.

In [3]:
%%sql 
CREATE TABLE IF NOT EXISTS customers
( 
    customer_id INTEGER PRIMARY KEY,
    customer_name TEXT NULL,
    gender TEXT NULL,
    age INTEGER NULL,
    home_address TEXT NULL,
    zip_code INTEGER NULL,
    city TEXT NULL,
    state TEXT NULL,
    country TEXT NULL
);

CREATE TABLE IF NOT EXISTS orders
(
    order_id INTEGER PRIMARY KEY,
    customer_id INTEGER REFERENCES customers (customer_id) NOT NULL,
    payment NUMERIC NULL,
    order_date DATE NULL,
    delivery_date DATE NULL
);

CREATE TABLE IF NOT EXISTS products
(
    product_id INTEGER PRIMARY KEY,
    product_type TEXT NULL,
    product_name TEXT NULL,
    size TEXT NULL,
    colour TEXT NULL,
    price NUMERIC NULL,
    quantity INTEGER NULL,
    description TEXT NULL
);

CREATE TABLE IF NOT EXISTS sales
(
    sales_id INTEGER PRIMARY KEY,
    order_id INTEGER REFERENCES orders (order_id) NOT NULL,
    product_id INTEGER REFERENCES products (product_id) NOT NULL,
    price_per_unit NUMERIC NULL,
    quantity INTEGER NULL,
    total_price NUMERIC NULL
);

COPY customers(customer_id,customer_name,gender,age,home_address,zip_code,city,state,country) 
FROM 'datasets/customers.csv' DELIMITER ',' CSV HEADER;

COPY orders(order_id,customer_id,payment,order_date,delivery_date) 
FROM 'datasets/orders.csv' DELIMITER ',' CSV HEADER;

COPY products(product_id,product_type,product_name,size,colour,price,quantity,description) 
FROM 'datasets/products.csv' DELIMITER ',' CSV HEADER;

COPY sales(sales_id,order_id,product_id,price_per_unit,quantity,total_price) 
FROM 'datasets/sales.csv' DELIMITER ',' CSV HEADER;


 * postgresql://postgres:***@localhost:5433/shopping_cart
Done.
Done.
Done.
Done.
1000 rows affected.
1000 rows affected.
1260 rows affected.
5000 rows affected.


[]

### Data exploration

Let's start the data analysis by having a quick overview of each table.

#### Customers table

In [4]:
%sql SELECT * FROM customers LIMIT 5

 * postgresql://postgres:***@localhost:5433/shopping_cart
5 rows affected.


[(1, 'Leanna Busson', 'Female', 30, '8606 Victoria TerraceSuite 560', 5464, 'Johnstonhaven', 'Northern Territory', 'Australia'),
 (2, 'Zabrina Harrowsmith', 'Genderfluid', 69, '8327 Kirlin SummitApt. 461', 8223, 'New Zacharyfort', 'South Australia', 'Australia'),
 (3, 'Shina Dullaghan', 'Polygender', 59, '269 Gemma SummitSuite 109', 5661, 'Aliburgh', 'Australian Capital Territory', 'Australia'),
 (4, 'Hewet McVitie', 'Bigender', 67, '743 Bailey GroveSuite 141', 1729, 'South Justinhaven', 'Queensland', 'Australia'),
 (5, 'Rubia Ashleigh', 'Polygender', 30, '48 Hyatt ManorSuite 375', 4032, 'Griffithsshire', 'Queensland', 'Australia')]

A first look to the table evidences that some customers can come from a same city or state. Genders are also non-unique. For example, one may identify the number of unique customers, genders and addresses.

In [5]:
%%sql 
SELECT 
    COUNT(DISTINCT customer_name) AS "# Customers",
    COUNT(DISTINCT gender) AS "# Genders",
    COUNT(DISTINCT home_address) AS "# Home Address",
    COUNT(DISTINCT city) AS "# Cities",
    COUNT(DISTINCT state) AS "# States",
    COUNT(DISTINCT country) AS "# Countries"
FROM customers

 * postgresql://postgres:***@localhost:5433/shopping_cart
1 rows affected.


[(1000, 8, 1000, 961, 8, 1)]

One can observe that there are 1000 customers, each with their own address among 961 cities and 8 states in Australia.

Next, the number and the age of customers can be determined for each state.

In [6]:
%%sql
SELECT 
    state AS "States",
    COUNT(customer_id) AS "# Customers" ,
    ROUND(COUNT(*)*100/SUM(COUNT(*)) OVER (),2) AS "% Customers"
FROM customers
GROUP BY state
ORDER BY state;

 * postgresql://postgres:***@localhost:5433/shopping_cart
8 rows affected.


[('Australian Capital Territory', 121, Decimal('12.10')),
 ('New South Wales', 132, Decimal('13.20')),
 ('Northern Territory', 125, Decimal('12.50')),
 ('Queensland', 134, Decimal('13.40')),
 ('South Australia', 139, Decimal('13.90')),
 ('Tasmania', 104, Decimal('10.40')),
 ('Victoria', 121, Decimal('12.10')),
 ('Western Australia', 124, Decimal('12.40'))]

In [7]:
%%sql
SELECT c.state,
    COUNT(CASE WHEN c.age < 20 THEN 1 END) AS "<20y",
    COUNT(CASE WHEN c.age BETWEEN 20 AND 40 THEN 1 END) AS "20-40y",
    COUNT(CASE WHEN c.age BETWEEN 40 AND 60 THEN 1 END) AS "40-60y",
    COUNT(CASE WHEN c.age BETWEEN 60 AND 80 THEN 1 END) AS "60-80y",
    COUNT(CASE WHEN c.age >= 80 THEN 1 END) AS "80y<"
FROM customers c
GROUP BY c.state
ORDER BY c.state

 * postgresql://postgres:***@localhost:5433/shopping_cart
8 rows affected.


[('Australian Capital Territory', 0, 37, 43, 42, 0),
 ('New South Wales', 0, 40, 41, 57, 3),
 ('Northern Territory', 0, 45, 49, 35, 1),
 ('Queensland', 0, 44, 46, 47, 4),
 ('South Australia', 0, 49, 47, 51, 2),
 ('Tasmania', 0, 37, 39, 35, 0),
 ('Victoria', 0, 46, 39, 38, 1),
 ('Western Australia', 0, 48, 38, 38, 4)]

The distribution is uniform over the states and ages ranging between 20 and 80 years old. One may also consider the gender of customers, showing also a uniform distribution.

In [8]:
%%sql
SELECT c.gender,
    COUNT(CASE WHEN c.age < 20 THEN 1 END) AS "<20y",
    COUNT(CASE WHEN c.age BETWEEN 20 AND 40 THEN 1 END) AS "20-40y",
    COUNT(CASE WHEN c.age BETWEEN 40 AND 60 THEN 1 END) AS "40-60y",
    COUNT(CASE WHEN c.age BETWEEN 60 AND 80 THEN 1 END) AS "60-80y",
    COUNT(CASE WHEN c.age >= 80 THEN 1 END) AS "80y<"
FROM customers c
GROUP BY c.gender
ORDER BY c.gender

 * postgresql://postgres:***@localhost:5433/shopping_cart
8 rows affected.


[('Agender', 0, 49, 36, 31, 2),
 ('Bigender', 0, 36, 46, 43, 1),
 ('Female', 0, 39, 47, 34, 0),
 ('Genderfluid', 0, 50, 35, 39, 2),
 ('Genderqueer', 0, 47, 35, 49, 1),
 ('Male', 0, 41, 52, 55, 4),
 ('Non-binary', 0, 42, 46, 47, 5),
 ('Polygender', 0, 42, 45, 45, 0)]

#### Sales table

In [9]:
%sql SELECT * FROM sales LIMIT 5

 * postgresql://postgres:***@localhost:5433/shopping_cart
5 rows affected.


[(0, 1, 218, Decimal('106'), 2, Decimal('212')),
 (1, 1, 481, Decimal('118'), 1, Decimal('118')),
 (2, 1, 2, Decimal('96'), 3, Decimal('288')),
 (3, 1, 1002, Decimal('106'), 2, Decimal('212')),
 (4, 1, 691, Decimal('113'), 3, Decimal('339'))]

In [10]:
%%sql
SELECT
    COUNT(sales_id) AS "# Sales",
    COUNT(DISTINCT order_id) AS "# Orders",
    COUNT(DISTINCT product_id) AS "# Products",
    MIN(price_per_unit) AS "Min. Price per Unit",
    MAX(price_per_unit) AS "Max. Price per Unit",
    MIN(quantity) AS "Min. Quantity",
    MAX(quantity) AS "Max. Quantity"
FROM sales

 * postgresql://postgres:***@localhost:5433/shopping_cart
1 rows affected.


[(5000, 993, 1233, Decimal('90'), Decimal('119'), 1, 3)]

For the <code>sales</code> table, one may observe that among the 5000 referenced sales, 993 orders and 1233 products are concerned. A maximum of 3 products are sold for each order, for which the price per unit is between 90 and 119 AUD.

Let's count now the number of sales according to the quantity of products sold. One may observe an uniform distribution.

In [11]:
%%sql
SELECT 
    quantity AS "Quantity of Products",
    COUNT(sales_id) AS "# Sales"
FROM sales
GROUP BY quantity
ORDER BY quantity DESC

 * postgresql://postgres:***@localhost:5433/shopping_cart
3 rows affected.


[(3, 1611), (2, 1740), (1, 1649)]

Now, the number of sales for the same products are estimated.

In [12]:
%%sql
WITH count_products AS (
    SELECT
        COUNT(sales_id) AS count_sales,
        product_id
    FROM sales
    GROUP BY product_id
)
SELECT
   count_sales AS "# Sales per Product",
   COUNT(product_id) AS "# Products"
FROM count_products
GROUP BY count_sales 
ORDER BY count_sales

 * postgresql://postgres:***@localhost:5433/shopping_cart
13 rows affected.


[(1, 88),
 (2, 168),
 (3, 254),
 (4, 264),
 (5, 210),
 (6, 128),
 (7, 63),
 (8, 38),
 (9, 12),
 (10, 4),
 (11, 2),
 (12, 1),
 (13, 1)]

One may notice that most of products are sold between 2 and 6 times.

#### Orders table

In [13]:
%sql SELECT * FROM orders LIMIT 5

 * postgresql://postgres:***@localhost:5433/shopping_cart
5 rows affected.


[(1, 64, Decimal('30811'), datetime.date(2021, 8, 30), datetime.date(2021, 9, 24)),
 (2, 473, Decimal('50490'), datetime.date(2021, 2, 3), datetime.date(2021, 2, 13)),
 (3, 774, Decimal('46763'), datetime.date(2021, 10, 8), datetime.date(2021, 11, 3)),
 (4, 433, Decimal('39782'), datetime.date(2021, 5, 6), datetime.date(2021, 5, 19)),
 (5, 441, Decimal('14719'), datetime.date(2021, 3, 23), datetime.date(2021, 3, 24))]

In [14]:
%%sql
SELECT 
    COUNT(order_id) AS "# Orders",
    COUNT(DISTINCT customer_id) AS "# Customers",
    COUNT(DISTINCT order_date) AS "# Order Dates",
    COUNT(DISTINCT delivery_date) AS "# Delivery Dates"
FROM orders

 * postgresql://postgres:***@localhost:5433/shopping_cart
1 rows affected.


[(1000, 617, 291, 305)]

For the <code>orders</code>, only 617 customers made 1000 orders.

In [15]:
%%sql
SELECT
    MIN(order_date) AS "Min. Order Date",
    MAX(order_date) AS "Max. Order Date",
    MIN(delivery_date) AS "Min. Delivery Date",
    MAX(delivery_date) AS "Max. Delivery Date",
    ROUND(AVG(delivery_date-order_date),2) AS "Avg. Delivery Duration",
    MAX(delivery_date-order_date) AS "Max. Delivery Duration",
    MIN(delivery_date-order_date) AS "Min. Delivery Duration"
FROM orders

 * postgresql://postgres:***@localhost:5433/shopping_cart
1 rows affected.


[(datetime.date(2021, 1, 1), datetime.date(2021, 10, 24), datetime.date(2021, 1, 3), datetime.date(2021, 11, 20), Decimal('14.08'), 27, 1)]

If we briefly analyze the order and delivery dates, all of them appear in 2021. Also, the minimum and maximum delivery duration are 1 and 27 days, respectively, with an average at 14 days.

#### Products table

In [16]:
%sql SELECT * FROM products LIMIT 5

 * postgresql://postgres:***@localhost:5433/shopping_cart
5 rows affected.


[(0, 'Shirt', 'Oxford Cloth', 'XS', 'red', Decimal('114'), 66, 'A red coloured, XS sized, Oxford Cloth Shirt'),
 (1, 'Shirt', 'Oxford Cloth', 'S', 'red', Decimal('114'), 53, 'A red coloured, S sized, Oxford Cloth Shirt'),
 (2, 'Shirt', 'Oxford Cloth', 'M', 'red', Decimal('114'), 54, 'A red coloured, M sized, Oxford Cloth Shirt'),
 (3, 'Shirt', 'Oxford Cloth', 'L', 'red', Decimal('114'), 69, 'A red coloured, L sized, Oxford Cloth Shirt'),
 (4, 'Shirt', 'Oxford Cloth', 'XL', 'red', Decimal('114'), 47, 'A red coloured, XL sized, Oxford Cloth Shirt')]

In [17]:
%%sql
SELECT 
    COUNT(product_id) AS "# Products",
    COUNT(DISTINCT product_type) AS "# Product types",
    COUNT(DISTINCT product_name) AS "# Product names",
    COUNT(DISTINCT size) AS "# Sizes",
    COUNT(DISTINCT colour) AS "# Colours",
    MAX(price) AS "Max. Price",
    MIN(price) AS "Min. Price",
    ROUND(AVG(price),2) AS "Avg. Price",
    MAX(quantity) AS "Max. Quantity",
    MIN(quantity) AS "Min. Quantity",
    ROUND(AVG(quantity),2) AS "Avg. Quantity"
FROM products

 * postgresql://postgres:***@localhost:5433/shopping_cart
1 rows affected.


[(1260, 3, 35, 5, 7, Decimal('119'), Decimal('90'), Decimal('105.81'), 80, 40, Decimal('60.15'))]

In [18]:
%%sql
SELECT
    product_type,
    COUNT(product_id) AS count_products
FROM products
GROUP BY product_type
ORDER BY product_type

 * postgresql://postgres:***@localhost:5433/shopping_cart
3 rows affected.


[('Jacket', 420), ('Shirt', 420), ('Trousers', 420)]

In [19]:
%%sql
SELECT
    size,
    COUNT(product_id) AS count_products
FROM products
GROUP BY size
ORDER BY size

 * postgresql://postgres:***@localhost:5433/shopping_cart
5 rows affected.


[('L', 252), ('M', 252), ('S', 252), ('XL', 252), ('XS', 252)]

In [20]:
%%sql
SELECT
    colour,
    COUNT(product_id) AS count_products
FROM products
GROUP BY colour
ORDER BY colour

 * postgresql://postgres:***@localhost:5433/shopping_cart
7 rows affected.


[('blue', 180),
 ('green', 180),
 ('indigo', 180),
 ('orange', 180),
 ('red', 180),
 ('violet', 180),
 ('yellow', 180)]

Briefly, the <code>products</code> table contains data on 1260 products of 3 types (jacket, shirt and trousers), 5 sizes (XS, S, M, L, XL) and 7 colours (blue, green, indigo, orange, red, violet and yellow). Their price is ranging from 90 to 190 AUD, and the quantity from 40 to 80.

### Which products were sold the most in the last month?

Here, we identify which products were sold the most, and target their name, their size and their colour.

We first create a <code>VIEW</code> which stores sales quantity and total price during the last month.

In [21]:
%%sql
CREATE VIEW sales_orders AS (
    SELECT sales.product_id AS product_id,
        sales.quantity AS quantity, 
        sales.total_price AS total_price, 
        orders.order_date AS order_date
    FROM sales
    INNER JOIN orders ON sales.order_id = orders.order_id
    WHERE order_date > date_trunc('day',( SELECT max(order_date) FROM orders) - interval '1' MONTH)
)

 * postgresql://postgres:***@localhost:5433/shopping_cart
Done.


[]

In [22]:
%%sql

SELECT DISTINCT products.product_name AS product_name,
    products.size AS size,
    products.colour AS colour,
    sales_orders.quantity AS quantity,
    sales_orders.order_date AS order_date,
    sales_orders.total_price AS total_price
FROM products
INNER JOIN sales_orders ON sales_orders.product_id = products.product_id
WHERE sales_orders.quantity = ( SELECT MAX(quantity) FROM sales )
ORDER BY quantity DESC, total_price DESC
LIMIT 20

 * postgresql://postgres:***@localhost:5433/shopping_cart
20 rows affected.


[('Trench Coat', 'M', 'indigo', 3, datetime.date(2021, 9, 25), Decimal('357')),
 ('Trench Coat', 'M', 'orange', 3, datetime.date(2021, 10, 17), Decimal('357')),
 ('Flannel', 'L', 'violet', 3, datetime.date(2021, 10, 22), Decimal('354')),
 ('Flannel', 'S', 'indigo', 3, datetime.date(2021, 10, 24), Decimal('354')),
 ('Flannel', 'S', 'red', 3, datetime.date(2021, 10, 23), Decimal('354')),
 ('Flannel', 'XL', 'yellow', 3, datetime.date(2021, 10, 13), Decimal('354')),
 ('Flannel', 'XS', 'green', 3, datetime.date(2021, 10, 8), Decimal('354')),
 ('Flannel', 'XS', 'indigo', 3, datetime.date(2021, 10, 15), Decimal('354')),
 ('Puffer', 'L', 'red', 3, datetime.date(2021, 10, 3), Decimal('354')),
 ('Puffer', 'M', 'green', 3, datetime.date(2021, 10, 10), Decimal('354')),
 ('Puffer', 'S', 'indigo', 3, datetime.date(2021, 10, 3), Decimal('354')),
 ('Puffer', 'S', 'violet', 3, datetime.date(2021, 10, 10), Decimal('354')),
 ('Puffer', 'S', 'yellow', 3, datetime.date(2021, 10, 22), Decimal('354')),
 ('Puffer', 'XL', 'red', 3, datetime.date(2021, 10, 21), Decimal('354')),
 ('Shearling', 'L', 'indigo', 3, datetime.date(2021, 10, 7), Decimal('351')),
 ('Shearling', 'L', 'yellow', 3, datetime.date(2021, 10, 13), Decimal('351')),
 ('Shearling', 'S', 'indigo', 3, datetime.date(2021, 10, 19), Decimal('351')),
 ('Shearling', 'S', 'red', 3, datetime.date(2021, 9, 26), Decimal('351')),
 ('Shearling', 'S', 'violet', 3, datetime.date(2021, 10, 3), Decimal('351')),
 ('Shearling', 'XL', 'orange', 3, datetime.date(2021, 10, 13), Decimal('351'))]

In [23]:
%%sql

SELECT DISTINCT products.product_name AS product_name,
    SUM(sales_orders.quantity) AS quantity,
    SUM(sales_orders.total_price) AS total_price
FROM products
INNER JOIN sales_orders ON sales_orders.product_id = products.product_id
WHERE sales_orders.quantity = ( SELECT MAX(quantity) FROM sales )
GROUP BY product_name
ORDER BY quantity DESC, total_price DESC
LIMIT 15

 * postgresql://postgres:***@localhost:5433/shopping_cart
15 rows affected.


[('Denim', 36, Decimal('3492')),
 ('Pleated', 30, Decimal('3150')),
 ('High-Waisted', 30, Decimal('2970')),
 ('Joggers', 30, Decimal('2790')),
 ('Cropped', 24, Decimal('2760')),
 ('Cuban Collar', 24, Decimal('2520')),
 ('Windbreaker', 24, Decimal('2448')),
 ('Drawstring', 24, Decimal('2304')),
 ('Mandarin Collar', 21, Decimal('2289')),
 ('Camp Collared', 21, Decimal('2058')),
 ('Cords', 21, Decimal('1974')),
 ('Linen', 21, Decimal('1911')),
 ('Flannel', 18, Decimal('2124')),
 ('Puffer', 18, Decimal('2124')),
 ('Shearling', 18, Decimal('2106'))]

In [24]:
%%sql

SELECT DISTINCT products.size AS size,
    SUM(sales_orders.quantity) AS quantity,
    SUM(sales_orders.total_price) AS total_price
FROM products
INNER JOIN sales_orders ON sales_orders.product_id = products.product_id
WHERE sales_orders.quantity = ( SELECT MAX(quantity) FROM sales )
GROUP BY size
ORDER BY quantity DESC, total_price DESC

 * postgresql://postgres:***@localhost:5433/shopping_cart
5 rows affected.


[('XS', 120, Decimal('12207')),
 ('L', 114, Decimal('11607')),
 ('XL', 111, Decimal('11283')),
 ('M', 105, Decimal('10971')),
 ('S', 102, Decimal('10833'))]

In [25]:
%%sql

SELECT DISTINCT products.colour AS colour,
    SUM(sales_orders.quantity) AS quantity,
    SUM(sales_orders.total_price) AS total_price
FROM products
INNER JOIN sales_orders ON sales_orders.product_id = products.product_id
WHERE sales_orders.quantity = ( SELECT MAX(quantity) FROM sales )
GROUP BY colour
ORDER BY quantity DESC, total_price DESC

 * postgresql://postgres:***@localhost:5433/shopping_cart
7 rows affected.


[('indigo', 96, Decimal('9981')),
 ('orange', 87, Decimal('8883')),
 ('yellow', 81, Decimal('8298')),
 ('blue', 81, Decimal('8253')),
 ('red', 75, Decimal('7776')),
 ('violet', 69, Decimal('7284')),
 ('green', 63, Decimal('6426'))]

In short, the main sold products during the last month possesses the following characteristics:
 - name: Denim (36 sales, total price at 3492 AUD), Pleated (30 sales, total price at 3150 AUD), High-Waisted (30 sales, total price at 2970 AUD) and Joggers (30 sales, total price at 2790 AUD)
 - size: XS (120 sales, total price at 12207 AUD), L (114 sales, total price at 11607 AUD) or XL (111 sales, total price at 11283 AUD)
 - colour: indigo (96 sales, total price at 9981 AUD), orange (87 sales, total price at 8883 AUD), yellow (81 sales, total price at 8298 AUD) or blue (81 sales, total price at 8253 AUD)

### How have sales changed over the past few quarters?

To answer this question, we need to calculate the variation in total prices and quantities.
To do so, a <code>VIEW</code> is created to store sums of these values aggregated for each quarter.

In [26]:
%%sql
CREATE VIEW quarters AS (
    SELECT 
        SUM(sales.total_price) AS total_price,
        SUM(sales.quantity) AS quantity,
        TO_CHAR(date_trunc('quarter', orders.order_date),'mm-yyyy') AS quarter
    FROM sales 
    INNER JOIN orders ON orders.order_id = sales.order_id
    GROUP BY quarter
)

 * postgresql://postgres:***@localhost:5433/shopping_cart
Done.


[]

In [27]:
%%sql 
SELECT total_price AS "Total Price",
       total_price - LAG(total_price) OVER (ORDER BY quarter ASC) AS "Total Price Variation",
       quantity AS "Total Quantity",
       quantity - LAG(quantity) OVER (ORDER BY quarter ASC) AS "Total Quantity Variation",
       quarter AS "Quarter"
FROM
    quarters
ORDER BY "Quarter" ASC

 * postgresql://postgres:***@localhost:5433/shopping_cart
4 rows affected.


[(Decimal('345777'), None, 3327, None, '01-2021'),
 (Decimal('284585'), Decimal('-61192'), 2755, -572, '04-2021'),
 (Decimal('317172'), Decimal('32587'), 3066, 311, '07-2021'),
 (Decimal('84266'), Decimal('-232906'), 814, -2252, '10-2021')]

One can observe that the largest number of sales appear during the first quarter (3327 for 345777 AUD), the second one being for the third quarter (3066 for 317172 AUD). For both second and fourth quarters, the variation is negative because of the significant decrease of sales, i.e. respectively 2755 (284585 AUD) and 814 (84266 AUD).

### What are the customer demographics and preferences?

We start by joining tables in order to access more easily to all data of interest.

In [28]:
%%sql
CREATE VIEW data AS 
SELECT
    oc.country AS country,
    oc.state AS state,
    oc.age AS age,
    oc.gender AS gender,
    p.product_name AS product,
    p.description AS description,
    p.size AS size,
    p.colour AS colour,
    p.product_type AS type,
    s.quantity AS quantity,
    s.total_price AS total_price
FROM sales s
INNER JOIN (
    SELECT
        o.order_id AS order_id,
        c.customer_name AS customer_name,
        c.country AS country,
        c.state AS state,
        c.age AS age,
        c.gender AS gender
    FROM orders o
    INNER JOIN customers c ON c.customer_id = o.customer_id) 
    oc ON oc.order_id = s.order_id
INNER JOIN products p ON p.product_id = s.product_id



 * postgresql://postgres:***@localhost:5433/shopping_cart
Done.


[]

In [29]:
%%sql
SELECT *
FROM data
LIMIT 5

 * postgresql://postgres:***@localhost:5433/shopping_cart
5 rows affected.


[('Australia', 'South Australia', 75, 'Genderfluid', 'Chambray', 'A orange coloured, L sized, Chambray Shirt', 'L', 'orange', 'Shirt', 2, Decimal('212')),
 ('Australia', 'South Australia', 75, 'Genderfluid', 'Puffer', 'A indigo coloured, S sized, Puffer Jacket', 'S', 'indigo', 'Jacket', 1, Decimal('118')),
 ('Australia', 'South Australia', 75, 'Genderfluid', 'Oxford Cloth', 'A red coloured, M sized, Oxford Cloth Shirt', 'M', 'red', 'Shirt', 3, Decimal('288')),
 ('Australia', 'South Australia', 75, 'Genderfluid', 'Wool', 'A blue coloured, M sized, Wool Trousers', 'M', 'blue', 'Trousers', 2, Decimal('212')),
 ('Australia', 'South Australia', 75, 'Genderfluid', 'Parka', 'A indigo coloured, S sized, Parka Jacket', 'S', 'indigo', 'Jacket', 3, Decimal('339'))]

Let's explore first the relationship between products and sales with the age of customers.

In [30]:
%%sql
SELECT type,
    SUM(CASE WHEN age BETWEEN 20 AND 40 THEN quantity END) AS "20-40y",
    SUM(CASE WHEN age BETWEEN 40 AND 60 THEN quantity END) AS "40-60y",
    SUM(CASE WHEN age BETWEEN 60 AND 80 THEN quantity END) AS "60-80y",
    SUM(CASE WHEN age >= 80 THEN quantity END) AS "80y<"
FROM data
GROUP BY type
ORDER BY "20-40y" DESC

 * postgresql://postgres:***@localhost:5433/shopping_cart
3 rows affected.


[('Trousers', 1284, 1117, 1075, 35),
 ('Shirt', 1201, 1168, 1000, 37),
 ('Jacket', 1185, 1178, 1076, 42)]

In general, the number of sales decreases with the age of customers. For the 20-40y category, trousers are the most sold products, but the less for 40-60y and 80y<. One may observe an exact opposite trend for Jacket. For the 60-80y people, they prefer trousers and jackets to shirts.

In [31]:
%%sql
SELECT 
    size,
    SUM(quantity) AS quantity
FROM data
GROUP BY size
ORDER BY quantity DESC

 * postgresql://postgres:***@localhost:5433/shopping_cart
5 rows affected.


[('M', 2036), ('L', 1998), ('S', 1992), ('XS', 1990), ('XL', 1946)]

In [32]:
%%sql
SELECT 
    size,
    SUM(CASE WHEN age BETWEEN 20 AND 40 THEN quantity END) AS "20-40y",
    SUM(CASE WHEN age BETWEEN 40 AND 60 THEN quantity END) AS "40-60y",
    SUM(CASE WHEN age BETWEEN 60 AND 80 THEN quantity END) AS "60-80y",
    SUM(CASE WHEN age >= 80 THEN quantity END) AS "80y<"
FROM data
GROUP BY size
ORDER BY "20-40y" DESC

 * postgresql://postgres:***@localhost:5433/shopping_cart
5 rows affected.


[('S', 755, 675, 618, 23),
 ('M', 751, 696, 647, 25),
 ('XS', 746, 687, 646, 15),
 ('XL', 713, 651, 637, 35),
 ('L', 705, 754, 603, 16)]

The largest quantity of sold products is reached by M sizes. If one inspects with regards to the age category, smallest sizes are prefered by 20-40y people, as for 60-80y people even if XL sizes are also significant. For 40-60y people, L sizes are more sold, while XL sizes are sligthly preferred by 80y<.

If we look to the sold quantity in function either of the state or the gender, results look uniforms without showing clear trends.

In [33]:
%%sql 
SELECT state,
    COUNT(CASE WHEN size = 'XS' THEN 1 END) AS xs,
    COUNT(CASE WHEN size = 'S' THEN 1 END) AS s,
    COUNT(CASE WHEN size = 'M' THEN 1 END) AS m,
    COUNT(CASE WHEN size = 'L' THEN 1 END) AS l,
    COUNT(CASE WHEN size = 'XL' THEN 1 END) AS xl,
    COUNT(size) AS count
FROM data
GROUP BY state
ORDER BY count DESC

 * postgresql://postgres:***@localhost:5433/shopping_cart
8 rows affected.


[('South Australia', 148, 157, 141, 145, 127, 718),
 ('Queensland', 136, 131, 142, 141, 137, 687),
 ('New South Wales', 138, 115, 121, 143, 130, 647),
 ('Australian Capital Territory', 119, 130, 141, 121, 127, 638),
 ('Western Australia', 133, 115, 132, 99, 125, 604),
 ('Victoria', 117, 134, 119, 125, 96, 591),
 ('Tasmania', 113, 115, 109, 123, 126, 586),
 ('Northern Territory', 97, 101, 117, 105, 109, 529)]

In [34]:
%%sql 
SELECT gender,
    COUNT(CASE WHEN size = 'XS' THEN 1 END) AS xs,
    COUNT(CASE WHEN size = 'S' THEN 1 END) AS s,
    COUNT(CASE WHEN size = 'M' THEN 1 END) AS m,
    COUNT(CASE WHEN size = 'L' THEN 1 END) AS l,
    COUNT(CASE WHEN size = 'XL' THEN 1 END) AS xl,
    COUNT(size) AS count
FROM data
GROUP BY gender
ORDER BY count DESC

 * postgresql://postgres:***@localhost:5433/shopping_cart
8 rows affected.


[('Female', 155, 125, 145, 153, 134, 712),
 ('Genderfluid', 112, 135, 139, 157, 144, 687),
 ('Male', 136, 152, 133, 124, 127, 672),
 ('Genderqueer', 134, 139, 131, 108, 135, 647),
 ('Polygender', 132, 122, 139, 127, 116, 636),
 ('Non-binary', 123, 131, 111, 121, 112, 598),
 ('Agender', 116, 107, 104, 97, 102, 526),
 ('Bigender', 93, 87, 120, 115, 107, 522)]

One may also study if some preferences appear for colours.

In [35]:
%%sql 
SELECT 
    colour,
    SUM(quantity) AS quantity
FROM data
GROUP BY colour
ORDER BY quantity DESC

 * postgresql://postgres:***@localhost:5433/shopping_cart
7 rows affected.


[('blue', 1528),
 ('violet', 1448),
 ('indigo', 1445),
 ('red', 1407),
 ('yellow', 1389),
 ('green', 1386),
 ('orange', 1359)]

In [36]:
%%sql
SELECT 
    colour,
    SUM(CASE WHEN age BETWEEN 20 AND 40 THEN quantity END) AS "20-40y",
    SUM(CASE WHEN age BETWEEN 40 AND 60 THEN quantity END) AS "40-60y",
    SUM(CASE WHEN age BETWEEN 60 AND 80 THEN quantity END) AS "60-80y",
    SUM(CASE WHEN age >= 80 THEN quantity END) AS "80y<"
FROM data
GROUP BY colour
ORDER BY "20-40y" DESC

 * postgresql://postgres:***@localhost:5433/shopping_cart
7 rows affected.


[('indigo', 567, 462, 454, 25),
 ('blue', 559, 528, 490, 11),
 ('red', 550, 490, 410, 9),
 ('yellow', 507, 511, 410, 18),
 ('orange', 504, 458, 444, 19),
 ('green', 495, 501, 452, 20),
 ('violet', 488, 513, 491, 12)]

In general, blue is the most prefered colour while orange is the less, but differences are not important, as observed by comparing age categories.

If one considers now the preference for gender or state, distributions look uniforms with few differences, without highlighting a color more than another.

In [37]:
%%sql 
SELECT state,
    COUNT(CASE WHEN colour = 'indigo' THEN 1 END) AS indigo,
    COUNT(CASE WHEN colour = 'orange' THEN 1 END) AS orange,
    COUNT(CASE WHEN colour = 'yellow' THEN 1 END) AS yellow,
    COUNT(CASE WHEN colour = 'blue' THEN 1 END) AS blue,
    COUNT(CASE WHEN colour = 'red' THEN 1 END) AS red,
    COUNT(CASE WHEN colour = 'violet' THEN 1 END) AS violet,
    COUNT(CASE WHEN colour = 'green' THEN 1 END) AS green,
    COUNT(size) AS count
FROM data
GROUP BY state
ORDER BY count DESC

 * postgresql://postgres:***@localhost:5433/shopping_cart
8 rows affected.


[('South Australia', 95, 115, 113, 107, 99, 98, 91, 718),
 ('Queensland', 104, 102, 99, 96, 85, 112, 89, 687),
 ('New South Wales', 93, 105, 77, 93, 92, 81, 106, 647),
 ('Australian Capital Territory', 87, 93, 90, 91, 75, 96, 106, 638),
 ('Western Australia', 93, 78, 83, 102, 92, 91, 65, 604),
 ('Victoria', 88, 65, 73, 89, 95, 100, 81, 591),
 ('Tasmania', 86, 73, 74, 92, 86, 98, 77, 586),
 ('Northern Territory', 80, 66, 81, 92, 68, 63, 79, 529)]

In [38]:
%%sql 
SELECT gender,
    COUNT(CASE WHEN colour = 'indigo' THEN 1 END) AS indigo,
    COUNT(CASE WHEN colour = 'orange' THEN 1 END) AS orange,
    COUNT(CASE WHEN colour = 'yellow' THEN 1 END) AS yellow,
    COUNT(CASE WHEN colour = 'blue' THEN 1 END) AS blue,
    COUNT(CASE WHEN colour = 'red' THEN 1 END) AS red,
    COUNT(CASE WHEN colour = 'violet' THEN 1 END) AS violet,
    COUNT(CASE WHEN colour = 'green' THEN 1 END) AS green,
    COUNT(size) AS count
FROM data
GROUP BY gender
ORDER BY count DESC

 * postgresql://postgres:***@localhost:5433/shopping_cart
8 rows affected.


[('Female', 104, 107, 89, 115, 100, 100, 97, 712),
 ('Genderfluid', 103, 91, 98, 89, 98, 114, 94, 687),
 ('Male', 93, 96, 83, 110, 92, 99, 99, 672),
 ('Genderqueer', 94, 76, 103, 105, 70, 104, 95, 647),
 ('Polygender', 90, 91, 93, 85, 89, 95, 93, 636),
 ('Non-binary', 95, 68, 82, 107, 83, 75, 88, 598),
 ('Agender', 73, 78, 66, 75, 93, 71, 70, 526),
 ('Bigender', 74, 90, 76, 76, 67, 81, 58, 522)]